content/departments/engineering/dev/tools/grpc.md

```diff
@@ -0,0 +1,76 @@
+
+- `99th percentile response time per method over 2m`
+- `90th percentile response time per method over 2m`
+- `75th percentile response time per method over 2m`
+
+#### GRPC "internal errors" metrics
+
+- `Client baseline error percentage per-method over 2m` - captures the percentage of errors that are related to grpc itself. For example, if the rpc method is called and later given a cancellation signal, then it is expected that the rpc method will fail. In this case, the error percentage will be high, but it is not a cause for concern because the failures are unrelated to grpc itself.
+- `Client baseline response codes rate per-method over 2m` - This gives you a generally idea of how often a status code is returned. For example, if you see a high percentage of `InvalidArgument` status codes, then it is likely there is something wrong with the request being sent to the grpc server.
+
+- `Client-observed gRPC internal error percentage per-method over 2m` - This capture the percentage of errors that we know originate from the grpc / protobuf libraries we use
+- `Client-observed gRPC internal error response code rate per-method over 2m` - This gives you a generally idea of how often a specific grpc status code is returned.
+
+Here are the posssible [grpc status codes](https://grpc.github.io/grpc/core/md_doc_statuscodes.html) that can be returned by a rpc method.
+
+#### More on internal errors
+
+The GRPC "internal errors" metrics flags specific types of errors that we can attribute to the grpc/protobuf libraries we use. Usually, these errors might hint at some configuration or application logic issues specific to grpc that may need further investigation.
+
+This errors are currently being captured through the use of a grpc feature called [Interceptors](https://github.com/grpc/grpc-go/blob/master/examples/features/interceptor/README.md).
+
+How can we determine if an error is coming from grpc-go?
+
+The go-grpc library has a tendency to prefix most gRPC client/server errors with "grpc: ...". Our new interceptors are designed to detect this string pattern.
+
+Some of the limitations of this approach are:
+
+- The error strings used by https://github.com/grpc/grpc-go could be modified in the future.
+- Not all errors from https://github.com/grpc/grpc-go are necessarily transmitted over the network.
+- It's possible for someone to inadvertently create an application-level error that starts with "grpc: ...".
+- Errors originating from the library may not always be the library's fault, such as when encountering an application bug like sending a 5GB symbols response (see [gRPC 4mb message size limit](#grpc-4mb-message-size-limit)).
+
+See the this [sourcegraph search](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/grpc/internalerrs/common.go?L160-166) to see the current list of errors that are being captured as "internal errors"
```

Comment:
This should be a permalink, as the line numbers can change as when the file is modified.